Bacalhau crashes when running docker containers if `--timeout` is reach

The current implementation of the Docker executor waits for the container to stop running, fetches the container, then fetches the logs for the container before parsing the logs.

The problem with this is if the `context.Context` times out before the container stops running, then `ContainerWait` call will return an error, fetching the container, fetching the logs will also then fail, and then parsing the logs will panic as it assumes the logs object to be populated.

Ideally, Bacalhau should return all of the logs up to that point or at least not crash.

```
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x18 pc=0x10182bd44]

goroutine 421 [running]:
github.com/filecoin-project/bacalhau/pkg/executor.writeOutputResult({0xc003e54b00, 0x72}, {{0x0, 0x0}, {0x1022174ac, 0x6}, 0x40000000, 0xc002764370, 0x800, 0xc002764380})
	pkg/executor/results.go:34 +0x84
github.com/filecoin-project/bacalhau/pkg/executor.WriteJobResults.func1()
	pkg/executor/results.go:113 +0x45
go.ptx.dk/multierrgroup.(*Group).Go.func1()
	go.ptx.dk/multierrgroup@v0.0.2/multierrgroup.go:50 +0x67
created by go.ptx.dk/multierrgroup.(*Group).Go
	go.ptx.dk/multierrgroup@v0.0.2/multierrgroup.go:48 +0x8d
```

Command to reproduce the error:
`bacalhau docker run ubuntu:22.04 --id-only --timeout=10 --wait=false -- sleep 9223372036`

--- a/Makefile
+++ b/Makefile
@@ -276,7 +276,7 @@ devstack:
 
 .PHONY: devstack-one
 devstack-one:
-	IGNORE_PORT_FILES=true PREDICTABLE_API_PORT=1 go run . devstack --requester-nodes 0 --compute-nodes 0 --hybrid-nodes 1
+	IGNORE_PID_AND_PORT_FILES=true PREDICTABLE_API_PORT=1 go run . devstack --requester-nodes 0 --compute-nodes 0 --hybrid-nodes 1
 
 .PHONY: devstack-100
 devstack-100:

--- a/pkg/docker/utils.go
+++ b/pkg/docker/utils.go
@@ -147,15 +147,16 @@ func FollowLogs(ctx context.Context, dockerClient *dockerclient.Client, nameOrID
 	go func() {
 		stdoutBuffer := bufio.NewWriter(stdoutWriter)
 		stderrBuffer := bufio.NewWriter(stderrWriter)
+		defer closer.CloseWithLogOnError("stderrWriter", stderrWriter)
+		defer closer.CloseWithLogOnError("stdoutWriter", stdoutWriter)
+		defer stderrBuffer.Flush()
+		defer stdoutBuffer.Flush()
+		defer closer.CloseWithLogOnError("logsReader", logsReader)
+
 		_, err = stdcopy.StdCopy(stdoutBuffer, stderrBuffer, logsReader)
 		if err != nil && !errors.Is(err, context.Canceled) {
-			log.Ctx(ctx).Error().Err(err).Msg("error reading container logs")
+			log.Ctx(ctx).Err(err).Msg("error reading container logs")
 		}
-		logsReader.Close()
-		stdoutBuffer.Flush()
-		stderrBuffer.Flush()
-		stdoutWriter.Close()
-		stderrWriter.Close()
 	}()
 
 	return stdoutReader, stderrReader, nil
@@ -175,23 +176,6 @@ func GetLogs(ctx context.Context, dockerClient *dockerclient.Client, nameOrID st
 		err = wg.Wait()
 	}
 	return
-}
-
-func StopContainer(ctx context.Context, dockerClient *dockerclient.Client, nameOrID string) error {
-	container, err := GetContainer(ctx, dockerClient, nameOrID)
-	if err != nil {
-		return err
-	}
-	if container == nil {
-		return nil
-	}
-	log.Ctx(ctx).Debug().Msgf("Container requested to stop: %s", container.ID)
-	timeout := time.Millisecond * 100
-	err = dockerClient.ContainerStop(ctx, container.ID, &timeout)
-	if err != nil {
-		return err
-	}
-	return nil
 }
 
 func RemoveContainer(ctx context.Context, dockerClient *dockerclient.Client, nameOrID string) error {

--- a/pkg/executor/docker/executor.go
+++ b/pkg/executor/docker/executor.go
@@ -6,6 +6,7 @@ import (
 	"os"
 	"path/filepath"
 	"strings"
+	"time"
 
 	"github.com/filecoin-project/bacalhau/pkg/compute/capacity"
 	"github.com/pkg/errors"
@@ -117,7 +118,7 @@ func (e *Executor) RunShard(
 		return executor.FailResult(err)
 	}
 
-	inputStorageSpecs := []model.StorageSpec{}
+	var inputStorageSpecs []model.StorageSpec
 	inputStorageSpecs = append(inputStorageSpecs, shard.Job.Spec.Contexts...)
 	inputStorageSpecs = append(inputStorageSpecs, shardStorageSpec...)
 
@@ -128,7 +129,7 @@ func (e *Executor) RunShard(
 
 	// the actual mounts we will give to the container
 	// these are paths for both input and output data
-	mounts := []mount.Mount{}
+	var mounts []mount.Mount
 	for spec, volumeMount := range inputVolumes {
 		if volumeMount.Type == storage.StorageVolumeConnectorBind {
 			log.Ctx(ctx).Trace().Msgf("Input Volume: %+v %+v", spec, volumeMount)
@@ -209,7 +210,7 @@ func (e *Executor) RunShard(
 		Tty:        false,
 		Env:        useEnv,
 		Entrypoint: shard.Job.Spec.Docker.Entrypoint,
-		Labels:     e.jobContainerLabels(&shard),
+		Labels:     e.jobContainerLabels(shard),
 		WorkingDir: shard.Job.Spec.Docker.WorkingDirectory,
 	}
 
@@ -273,6 +274,9 @@ func (e *Executor) RunShard(
 		return executor.FailResult(internalContainerStartError)
 	}
 
+	log.Ctx(ctx).Debug().Msg("Capturing stdout/stderr for container")
+	stdoutPipe, stderrPipe, logsErr := docker.FollowLogs(ctx, e.Client, jobContainer.ID)
+
 	// the idea here is even if the container errors
 	// we want to capture stdout, stderr and feed it back to the user
 	var containerError error
@@ -292,9 +296,6 @@ func (e *Executor) RunShard(
 		}
 	}
 
-	log.Ctx(ctx).Debug().Msg("Capturing stdout/stderr for container")
-	stdoutPipe, stderrPipe, logsErr := docker.FollowLogs(ctx, e.Client, jobContainer.ID)
-
 	return executor.WriteJobResults(
 		jobResultsDir,
 		stdoutPipe,
@@ -305,7 +306,7 @@ func (e *Executor) RunShard(
 }
 
 func (e *Executor) CancelShard(ctx context.Context, shard model.JobShard) error {
-	return docker.StopContainer(ctx, e.Client, e.jobContainerName(shard))
+	return docker.RemoveObjectsWithLabel(ctx, e.Client, labelJobName, e.labelJobValue(shard))
 }
 
 func (e *Executor) cleanupJob(ctx context.Context, shard model.JobShard) {
@@ -313,7 +314,10 @@ func (e *Executor) cleanupJob(ctx context.Context, shard model.JobShard) {
 		return
 	}
 
-	err := docker.RemoveObjectsWithLabel(ctx, e.Client, labelJobName, e.ID+shard.ID())
+	// Use a separate context in case the current one has already been canceled
+	separateCtx, cancel := context.WithTimeout(context.Background(), 1*time.Minute)
+	defer cancel()
+	err := docker.RemoveObjectsWithLabel(separateCtx, e.Client, labelJobName, e.labelJobValue(shard))
 	logLevel := map[bool]zerolog.Level{true: zerolog.DebugLevel, false: zerolog.ErrorLevel}[err == nil]
 	log.Ctx(ctx).WithLevel(logLevel).Err(err).Msg("Cleaned up job Docker resources")
 }
@@ -342,11 +346,15 @@ func (e *Executor) jobContainerName(shard model.JobShard) string {
 	return e.dockerObjectName(shard, "executor")
 }
 
-func (e *Executor) jobContainerLabels(shard *model.JobShard) map[string]string {
+func (e *Executor) jobContainerLabels(shard model.JobShard) map[string]string {
 	return map[string]string{
 		labelExecutorName: e.ID,
-		labelJobName:      e.ID + shard.ID(),
+		labelJobName:      e.labelJobValue(shard),
 	}
+}
+
+func (e *Executor) labelJobValue(shard model.JobShard) string {
+	return e.ID + shard.ID()
 }
 
 // Compile-time interface check:

--- a/pkg/executor/docker/executor_test.go
+++ b/pkg/executor/docker/executor_test.go
@@ -12,6 +12,7 @@ import (
 	"strconv"
 	"strings"
 	"testing"
+	"time"
 
 	"github.com/filecoin-project/bacalhau/pkg/compute/capacity"
 	"github.com/filecoin-project/bacalhau/pkg/docker"
@@ -33,20 +34,20 @@ func TestExecutorTestSuite(t *testing.T) {
 	suite.Run(t, new(ExecutorTestSuite))
 }
 
-func (suite *ExecutorTestSuite) SetupTest() {
-	docker.MustHaveDocker(suite.T())
+func (s *ExecutorTestSuite) SetupTest() {
+	docker.MustHaveDocker(s.T())
 
 	var err error
-	suite.cm = system.NewCleanupManager()
-	suite.T().Cleanup(suite.cm.Cleanup)
+	s.cm = system.NewCleanupManager()
+	s.T().Cleanup(s.cm.Cleanup)
 
-	suite.executor, err = NewExecutor(
+	s.executor, err = NewExecutor(
 		context.Background(),
-		suite.cm,
+		s.cm,
 		"bacalhau-executor-unittest",
 		storage.NewMappedStorageProvider(map[model.StorageSourceType]storage.Storage{}),
 	)
-	require.NoError(suite.T(), err)
+	require.NoError(s.T(), err)
 
 	handler := func(w http.ResponseWriter, r *http.Request) {
 		w.Write([]byte(r.URL.Path))
@@ -59,28 +60,28 @@ func (suite *ExecutorTestSuite) SetupTest() {
 	// the "docker0" interface.
 	var gateway net.IP
 	if runtime.GOOS == "linux" {
-		gateway, err = docker.HostGatewayIP(context.Background(), suite.executor.Client)
-		require.NoError(suite.T(), err)
+		gateway, err = docker.HostGatewayIP(context.Background(), s.executor.Client)
+		require.NoError(s.T(), err)
 	} else {
 		gateway = net.ParseIP("127.0.0.1")
 	}
 
 	serverAddr := net.TCPAddr{IP: gateway, Port: 0}
 	listener, err := net.Listen("tcp", serverAddr.String())
-	require.NoError(suite.T(), err)
+	require.NoError(s.T(), err)
 	// Don't need to close the listener as it'll be closed by the server.
 
-	suite.server = &http.Server{
+	s.server = &http.Server{
 		Addr:    listener.Addr().String(),
 		Handler: http.HandlerFunc(handler),
 	}
-	suite.cm.RegisterCallback(suite.server.Close)
-	go suite.server.Serve(listener)
+	s.cm.RegisterCallback(s.server.Close)
+	go s.server.Serve(listener)
 }
 
-func (suite *ExecutorTestSuite) containerHttpURL() *url.URL {
-	url, err := url.Parse("http://" + suite.server.Addr)
-	require.NoError(suite.T(), err)
+func (s *ExecutorTestSuite) containerHttpURL() *url.URL {
+	url, err := url.Parse("http://" + s.server.Addr)
+	require.NoError(s.T(), err)
 
 	// On Mac/Windows, we are within a VM and hence we need to route to the
 	// host. On Linux we are not, so localhost should work.
@@ -89,22 +90,26 @@ func (suite *ExecutorTestSuite) containerHttpURL() *url.URL {
 	return url
 }
 
-func (suite *ExecutorTestSuite) curlTask() model.JobSpecDocker {
+func (s *ExecutorTestSuite) curlTask() model.JobSpecDocker {
 	return model.JobSpecDocker{
 		Image:      "curlimages/curl",
-		Entrypoint: []string{"curl", "--fail-with-body", suite.containerHttpURL().JoinPath("hello.txt").String()},
+		Entrypoint: []string{"curl", "--fail-with-body", s.containerHttpURL().JoinPath("hello.txt").String()},
 	}
 }
 
-func (suite *ExecutorTestSuite) runJob(spec model.Spec) (*model.RunCommandResult, error) {
-	result := suite.T().TempDir()
-	j := &model.Job{Metadata: model.Metadata{ID: "test"}, Spec: spec}
-	shard := model.JobShard{Job: j, Index: 0}
-	return suite.executor.RunShard(context.Background(), shard, result)
+func (s *ExecutorTestSuite) runJob(spec model.Spec) (*model.RunCommandResult, error) {
+	return s.runJobWithContext(context.Background(), spec)
 }
 
-func (suite *ExecutorTestSuite) runJobGetStdout(spec model.Spec) (string, error) {
-	runnerOutput, runErr := suite.runJob(spec)
+func (s *ExecutorTestSuite) runJobWithContext(ctx context.Context, spec model.Spec) (*model.RunCommandResult, error) {
+	result := s.T().TempDir()
+	j := &model.Job{Metadata: model.Metadata{ID: "test"}, Spec: spec}
+	shard := model.JobShard{Job: j, Index: 0}
+	return s.executor.RunShard(ctx, shard, result)
+}
+
+func (s *ExecutorTestSuite) runJobGetStdout(spec model.Spec) (string, error) {
+	runnerOutput, runErr := s.runJob(spec)
 	return runnerOutput.STDOUT, runErr
 }
 
@@ -113,15 +118,15 @@ const (
 	MEMORY_LIMIT = "100mb"
 )
 
-func (suite *ExecutorTestSuite) TestDockerResourceLimitsCPU() {
+func (s *ExecutorTestSuite) TestDockerResourceLimitsCPU() {
 	if runtime.GOOS == "windows" {
-		suite.T().Skip("Resource limits don't apply to containers running on Windows")
+		s.T().Skip("Resource limits don't apply to containers running on Windows")
 	}
 
 	// this will give us a numerator and denominator that should end up at the
 	// same 0.1 value that 100m means
 	// https://access.redhat.com/documentation/en-us/red_hat_enterprise_linux/8/html/managing_monitoring_and_updating_the_kernel/using-cgroups-v2-to-control-distribution-of-cpu-time-for-applications_managing-monitoring-and-updating-the-kernel#proc_controlling-distribution-of-cpu-time-for-applications-by-adjusting-cpu-bandwidth_using-cgroups-v2-to-control-distribution-of-cpu-time-for-applications
-	result, err := suite.runJobGetStdout(model.Spec{
+	result, err := s.runJobGetStdout(model.Spec{
 		Engine: model.EngineDocker,
 		Resources: model.ResourceUsageConfig{
 			CPU:    CPU_LIMIT,
@@ -136,10 +141,10 @@ func (suite *ExecutorTestSuite) TestDockerResourceLimitsCPU() {
 	values := strings.Fields(result)
 
 	numerator, err := strconv.Atoi(values[0])
-	require.NoError(suite.T(), err)
+	require.NoError(s.T(), err)
 
 	denominator, err := strconv.Atoi(values[1])
-	require.NoError(suite.T(), err)
+	require.NoError(s.T(), err)
 
 	var containerCPU float64 = 0
 
@@ -147,18 +152,18 @@ func (suite *ExecutorTestSuite) TestDockerResourceLimitsCPU() {
 		containerCPU = float64(numerator) / float64(denominator)
 	}
 
-	require.Equal(suite.T(), capacity.ConvertCPUString(CPU_LIMIT), containerCPU, "the container reported CPU does not equal the configured limit")
+	require.Equal(s.T(), capacity.ConvertCPUString(CPU_LIMIT), containerCPU, "the container reported CPU does not equal the configured limit")
 }
 
-func (suite *ExecutorTestSuite) TestDockerResourceLimitsMemory() {
+func (s *ExecutorTestSuite) TestDockerResourceLimitsMemory() {
 	if runtime.GOOS == "windows" {
-		suite.T().Skip("Resource limits don't apply to containers running on Windows")
+		s.T().Skip("Resource limits don't apply to containers running on Windows")
 	}
 
 	// this will give us a numerator and denominator that should end up at the
 	// same 0.1 value that 100m means
 	// https://access.redhat.com/documentation/en-us/red_hat_enterprise_linux/8/html/managing_monitoring_and_updating_the_kernel/using-cgroups-v2-to-control-distribution-of-cpu-time-for-applications_managing-monitoring-and-updating-the-kernel#proc_controlling-distribution-of-cpu-time-for-applications-by-adjusting-cpu-bandwidth_using-cgroups-v2-to-control-distribution-of-cpu-time-for-applications
-	result, err := suite.runJobGetStdout(model.Spec{
+	result, err := s.runJobGetStdout(model.Spec{
 		Engine: model.EngineDocker,
 		Resources: model.ResourceUsageConfig{
 			CPU:    CPU_LIMIT,
@@ -169,88 +174,88 @@ func (suite *ExecutorTestSuite) TestDockerResourceLimitsMemory() {
 			Entrypoint: []string{"bash", "-c", "cat /sys/fs/cgroup/memory.max"},
 		},
 	})
-	require.NoError(suite.T(), err)
+	require.NoError(s.T(), err)
 
 	intVar, err := strconv.Atoi(strings.TrimSpace(result))
-	require.NoError(suite.T(), err)
-	require.Equal(suite.T(), capacity.ConvertBytesString(MEMORY_LIMIT), uint64(intVar), "the container reported memory does not equal the configured limit")
+	require.NoError(s.T(), err)
+	require.Equal(s.T(), capacity.ConvertBytesString(MEMORY_LIMIT), uint64(intVar), "the container reported memory does not equal the configured limit")
 }
 
-func (suite *ExecutorTestSuite) TestDockerNetworkingFull() {
-	result, err := suite.runJob(model.Spec{
+func (s *ExecutorTestSuite) TestDockerNetworkingFull() {
+	result, err := s.runJob(model.Spec{
 		Engine:  model.EngineDocker,
 		Network: model.NetworkConfig{Type: model.NetworkFull},
-		Docker:  suite.curlTask(),
+		Docker:  s.curlTask(),
 	})
-	require.NoError(suite.T(), err, result.STDERR)
-	require.Zero(suite.T(), result.ExitCode, result.STDERR)
-	require.Equal(suite.T(), "/hello.txt", result.STDOUT)
+	require.NoError(s.T(), err, result.STDERR)
+	require.Zero(s.T(), result.ExitCode, result.STDERR)
+	require.Equal(s.T(), "/hello.txt", result.STDOUT)
 }
 
-func (suite *ExecutorTestSuite) TestDockerNetworkingNone() {
-	result, err := suite.runJob(model.Spec{
+func (s *ExecutorTestSuite) TestDockerNetworkingNone() {
+	result, err := s.runJob(model.Spec{
 		Engine:  model.EngineDocker,
 		Network: model.NetworkConfig{Type: model.NetworkNone},
-		Docker:  suite.curlTask(),
+		Docker:  s.curlTask(),
 	})
-	require.NoError(suite.T(), err)
-	require.Empty(suite.T(), result.STDOUT)
-	require.NotEmpty(suite.T(), result.STDERR)
-	require.NotZero(suite.T(), result.ExitCode)
+	require.NoError(s.T(), err)
+	require.Empty(s.T(), result.STDOUT)
+	require.NotEmpty(s.T(), result.STDERR)
+	require.NotZero(s.T(), result.ExitCode)
 }
 
-func (suite *ExecutorTestSuite) TestDockerNetworkingHTTP() {
-	result, err := suite.runJob(model.Spec{
+func (s *ExecutorTestSuite) TestDockerNetworkingHTTP() {
+	result, err := s.runJob(model.Spec{
 		Engine: model.EngineDocker,
 		Network: model.NetworkConfig{
 			Type:    model.NetworkHTTP,
-			Domains: []string{suite.containerHttpURL().Hostname()},
+			Domains: []string{s.containerHttpURL().Hostname()},
 		},
-		Docker: suite.curlTask(),
+		Docker: s.curlTask(),
 	})
-	require.NoError(suite.T(), err, result.STDERR)
-	require.Zero(suite.T(), result.ExitCode, result.STDERR)
-	require.Equal(suite.T(), "/hello.txt", result.STDOUT)
+	require.NoError(s.T(), err, result.STDERR)
+	require.Zero(s.T(), result.ExitCode, result.STDERR)
+	require.Equal(s.T(), "/hello.txt", result.STDOUT)
 }
 
-func (suite *ExecutorTestSuite) TestDockerNetworkingHTTPWithMultipleDomains() {
-	result, err := suite.runJob(model.Spec{
+func (s *ExecutorTestSuite) TestDockerNetworkingHTTPWithMultipleDomains() {
+	result, err := s.runJob(model.Spec{
 		Engine: model.EngineDocker,
 		Network: model.NetworkConfig{
 			Type: model.NetworkHTTP,
 			Domains: []string{
-				suite.containerHttpURL().Hostname(),
+				s.containerHttpURL().Hostname(),
 				"bacalhau.org",
 			},
 		},
-		Docker: suite.curlTask(),
+		Docker: s.curlTask(),
 	})
-	require.NoError(suite.T(), err, result.STDERR)
-	require.Zero(suite.T(), result.ExitCode, result.STDERR)
-	require.Equal(suite.T(), "/hello.txt", result.STDOUT)
+	require.NoError(s.T(), err, result.STDERR)
+	require.Zero(s.T(), result.ExitCode, result.STDERR)
+	require.Equal(s.T(), "/hello.txt", result.STDOUT)
 }
 
-func (suite *ExecutorTestSuite) TestDockerNetworkingFiltersHTTP() {
-	result, err := suite.runJob(model.Spec{
+func (s *ExecutorTestSuite) TestDockerNetworkingFiltersHTTP() {
+	result, err := s.runJob(model.Spec{
 		Engine: model.EngineDocker,
 		Network: model.NetworkConfig{
 			Type:    model.NetworkHTTP,
 			Domains: []string{"bacalhau.org"},
 		},
-		Docker: suite.curlTask(),
+		Docker: s.curlTask(),
 	})
 	// The curl will succeed but should return a non-zero exit code and error page.
-	require.NoError(suite.T(), err)
-	require.NotZero(suite.T(), result.ExitCode)
-	require.Contains(suite.T(), result.STDOUT, "ERROR: The requested URL could not be retrieved")
+	require.NoError(s.T(), err)
+	require.NotZero(s.T(), result.ExitCode)
+	require.Contains(s.T(), result.STDOUT, "ERROR: The requested URL could not be retrieved")
 }
 
-func (suite *ExecutorTestSuite) TestDockerNetworkingFiltersHTTPS() {
-	result, err := suite.runJob(model.Spec{
+func (s *ExecutorTestSuite) TestDockerNetworkingFiltersHTTPS() {
+	result, err := s.runJob(model.Spec{
 		Engine: model.EngineDocker,
 		Network: model.NetworkConfig{
 			Type:    model.NetworkHTTP,
-			Domains: []string{suite.containerHttpURL().Hostname()},
+			Domains: []string{s.containerHttpURL().Hostname()},
 		},
 		Docker: model.JobSpecDocker{
 			Image:      "curlimages/curl",
@@ -258,23 +263,40 @@ func (suite *ExecutorTestSuite) TestDockerNetworkingFiltersHTTPS() {
 		},
 	})
 	// The curl will succeed but should return a non-zero exit code and error page.
-	require.NoError(suite.T(), err)
-	require.NotZero(suite.T(), result.ExitCode)
-	require.Empty(suite.T(), result.STDOUT)
+	require.NoError(s.T(), err)
+	require.NotZero(s.T(), result.ExitCode)
+	require.Empty(s.T(), result.STDOUT)
 }
 
-func (suite *ExecutorTestSuite) TestDockerNetworkingAppendsHTTPHeader() {
-	suite.server.Handler = http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
-		w.Write([]byte(r.Header.Get("X-Bacalhau-Job-ID")))
+func (s *ExecutorTestSuite) TestDockerNetworkingAppendsHTTPHeader() {
+	s.server.Handler = http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		_, err := w.Write([]byte(r.Header.Get("X-Bacalhau-Job-ID")))
+		s.Require().NoError(err)
 	})
-	result, err := suite.runJob(model.Spec{
+	result, err := s.runJob(model.Spec{
 		Engine: model.EngineDocker,
 		Network: model.NetworkConfig{
 			Type:    model.NetworkHTTP,
-			Domains: []string{suite.containerHttpURL().Hostname()},
+			Domains: []string{s.containerHttpURL().Hostname()},
 		},
-		Docker: suite.curlTask(),
+		Docker: s.curlTask(),
 	})
-	require.NoError(suite.T(), err)
-	require.Equal(suite.T(), "test", result.STDOUT, result.STDOUT)
+	require.NoError(s.T(), err)
+	require.Equal(s.T(), "test", result.STDOUT, result.STDOUT)
+}
+
+func (s *ExecutorTestSuite) TestTimesOutCorrectly() {
+	expected := "message after sleep"
+	ctx, cancel := context.WithTimeout(context.Background(), 2*time.Second)
+	defer cancel()
+
+	result, err := s.runJobWithContext(ctx, model.Spec{
+		Engine: model.EngineDocker,
+		Docker: model.JobSpecDocker{
+			Image:      "ubuntu",
+			Entrypoint: []string{"bash", "-c", fmt.Sprintf(`sleep 1 && echo "%s" && sleep 20`, expected)},
+		},
+	})
+	s.ErrorIs(err, context.DeadlineExceeded)
+	s.Truef(strings.HasPrefix(result.STDOUT, expected), "'%s' does not start with '%s'", result.STDOUT, expected)
 }

--- a/pkg/executor/docker/network.go
+++ b/pkg/executor/docker/network.go
@@ -106,7 +106,7 @@ func (e *Executor) createHTTPGateway(
 		Scope:      "local",
 		Internal:   true,
 		Attachable: true,
-		Labels:     e.jobContainerLabels(&shard),
+		Labels:     e.jobContainerLabels(shard),
 	})
 	if err != nil {
 		return nil, nil, errors.Wrap(err, "error creating network")
@@ -135,7 +135,7 @@ func (e *Executor) createHTTPGateway(
 		},
 		Healthcheck:     &container.HealthConfig{}, //TODO
 		NetworkDisabled: false,
-		Labels:          e.jobContainerLabels(&shard),
+		Labels:          e.jobContainerLabels(shard),
 	}, &container.HostConfig{
 		NetworkMode: dockerNetworkBridge,
 		CapAdd:      gatewayCapabilities,

--- a/pkg/executor/results.go
+++ b/pkg/executor/results.go
@@ -11,6 +11,7 @@ import (
 	"github.com/filecoin-project/bacalhau/pkg/ipfs"
 	"github.com/filecoin-project/bacalhau/pkg/model"
 	"github.com/filecoin-project/bacalhau/pkg/system"
+	"github.com/filecoin-project/bacalhau/pkg/util/closer"
 	"go.ptx.dk/multierrgroup"
 	"go.uber.org/multierr"
 )
@@ -49,7 +50,7 @@ func writeOutputResult(resultsDir string, output outputResult) error {
 	if err != nil {
 		return err
 	}
-	defer file.Close()
+	defer closer.CloseWithLogOnError("file", file)
 
 	// First write the bytes we have already read, and then write whatever
 	// is left in the buffer, but only up to the maximum file limit.
@@ -67,7 +68,7 @@ func writeOutputResult(resultsDir string, output outputResult) error {
 	return nil
 }
 
-// WriteJobResult produces files and a model.RunCommandResult in the standard
+// WriteJobResults produces files and a model.RunCommandResult in the standard
 // format, including truncating the contents of both where necessary to fit
 // within system-defined limits.
 //

